design/backup-performance-improvements.md

```diff
@@ -0,0 +1,358 @@
+`SetupWithManager` will also add the input channel to the `itemBackupper` so that it will be available during backup processing.
+The func `RunItemBlockWorkers` will create the `ItemBlockWorkerPool` with a shared buffered input channel (fixed buffer size) and start `workers` gororoutines which will each call `processItemBlocksWorker`.
+The `processItemBlocksWorker` func (run by the worker goroutines) will read from `itemBlockChannel`, call `BackupItemBlock` on the retrieved `ItemBlock`, and then send the return value to the retrieved `returnChan`, and then process the next block.
+
+#### Modify ItemBlock processing loop to send ItemBlocks to the worker pool rather than backing them up directly
+
+The ItemBlock processing loop implemented in Phase 1 will be modified to send each newly-created ItemBlock to the shared channel rather than calling `BackupItemBlock` inline, using a WaitGroup to manage in-process items. A separate goroutine will be created to process returns for this backup. After completion of the ItemBlock processing loop, velero will use the WaitGroup to wait for all ItemBlock processing to complete before moving forward.
+
+A simplified example of what this response goroutine might look like:
+```
+    // omitting cancel handling, context, etc
+    ret := make(chan ItemBlockReturn)
+    wg := &sync.WaitGroup{}
+    // Handle returns
+    go func() {
+        for {
+            select {
+            case response := <-ret: // process each BackupItemBlock response
+                func() {
+                    defer wg.Done()
+                    responses = append(responses, response)
+                }()
+            case <-ctx.Done():
+                return
+            }
+        }
+    }()
+    // Simplified illustration, looping over and assumed already-determined ItemBlock list
+    for _, itemBlock := range itemBlocks {
+        wg.Add(1)
+        inputChan <- ItemBlockInput{itemBlock: itemBlock, returnChan: ret}
+    }
+    done := make(chan struct{})
+    go func() {
+        defer close(done)
+        wg.Wait()
+    }()
+    // Wait for all the ItemBlocks to be processed
+    select {
+    case <-done:
+        logger.Info("done processing ItemBlocks")
+    }
+    // responses from BackupItemBlock calls are in responses
+```
+
+The ItemBlock processing loop described above will be split into two separate iterations. For the first iteration, velero will only process those items at the beginning of the loop identified as `orderedResources` -- when the groups generated from these resources are passed to the worker channel, velero will wait for the response before moving on to the next ItemBlock. This is to ensure that the ordered resources are processed in the required order. Once the last ordered resource is processed, the remaining ItemBlocks will be processed and sent to the worker channel without waiting for a response, in order to allow these ItemBlocks to be processed in parallel.
```

Comment:
Does it mean the items identified as `orderedResources` are always processed sequentially?